### PR TITLE
Add new RET_NERRNO() helper

### PR DESCRIPTION
**Pull request by poettering**
Result of this discussion:

https://github.com/systemd/systemd/pull/21326#discussion_r748413537

I kinda like the result. 

I went for the name RET_NERRNO() which emphasizes that this is about fixing up return values, and that we pick up `errno` for this, but negate it.

### CODE REVIEW

**Comment by poettering** [comment]
kinda nice: +172 −467 → i.e. almost 300 net lines removed. nice.

**Review by yuwata — APPROVED** [review]
LGTM, though I only checked the first few changes in the third commit.

**Review by bluca — APPROVED** [review]
lgtm - just to double check, the `ioctl` manpage says:

> Usually,  on  success  zero is returned.  A few ioctl() requests use the return value as an output parameter and return a nonnegative value on success.

Are we sure all these cases where the manual `return 0` was removed are fine?

**Comment by DaanDeMeyer** [comment]
Perhaps also document this in the coding style doc?

**Comment by poettering** [comment]
> Are we sure all these cases where the manual `return 0` was removed are fine?

I tried to make sure of this, yes. i.e. I looked individually at all ioctls i changed, and did not do a blanket replacement. In our codebase I found very few ioctls that return anything else but zero (NS_GET_NSTYPE is one), but that one wasn't even a candidate for this new macro...

So I think we are fine (but of course, reviews regarding this welcome)

**Comment by poettering** [comment]
> Perhaps also document this in the coding style doc?

Added. PTAL!

File: src/basic/errno-util.h [hunk 1]
```diff
@@ -31,6 +31,29 @@ static inline int negative_errno(void) {
...
+         *
+         * or
+         *
+         *     …
+         *     fd = RET_NERRNO(open("/etc/fstab", O_RDONLY|O_CLOEXEC));
+         *     …
+         */
+
+        if (ret < 0)
+                return negative_errno();
```
Comment (keszybz) [inline src/basic/errno-util.h:52]: Note to self: `negative_errno` already emits a warning if errno isn't set properly. We haven't seen any of those, so it seems errno is really set everywhere it should be.

File: src/core/execute.c [hunk 2]
```diff
@@ -641,7 +641,7 @@ static int setup_output(
 
                 /* Duplicate from stdout if possible */
                 if (can_inherit_stderr_from_stdout(context, o, e))
-                        return dup2(STDOUT_FILENO, fileno) < 0 ? -errno : fileno;
+                        return RET_NERRNO(dup2(STDOUT_FILENO, fileno));
```
Comment (keszybz) [inline src/core/execute.c:644]: Note to self: dup2 always returns the second arg or error, so that change is OK.

**Review by keszybz — APPROVED** [review]
Looks great.

**Pull request merged**